### PR TITLE
Add create shopcart

### PR DESCRIPTION
**Pull request by KHafner**
# Pull Request Template

## Checklist:

- [x] I have written test cases for this change, and my code passed the test
- [x] I have updated documentation to reflect this change
- [x] I have attached this PR to an issue/user story

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -62,3 +62,6 @@ You can also manually run `nosetests` with `coverage` (but `setup.cfg` does this
 ## Logging Services
 
 Logging is set up to track events.
+
+## Creating a Shopcart or Shopcart Item
+A shopcart or shopcart item can be created with a `POST` request on `'/shopcarts'` or '`/shopcartitems'` respectively.
```
Comment (iLtc) [inline README.md:67]: Could you list the required parameters / JSON format here?

File: service/models.py [hunk 2]
```diff
@@ -63,22 +64,56 @@ class Shopcart(db.Model):
...
+        """
+        Creates a Shopcart to the database
+        """
+        date_time = datetime.now()
+        self.id = None  # id must be none to generate next primary key
+        #if time fields not set, use current time
+        if self.create_time is None:
+            self.create_time = date_time
+        if self.update_time is None:
+            self.update_time = date_time
```
Comment (iLtc) [inline service/models.py:87]: Just curious, if we already have `server_default` and `onupdate`, do we need to set the value here?

File: service/models.py [hunk 3]
```diff
@@ -63,22 +64,56 @@ class Shopcart(db.Model):
...
+        if isinstance(self.update_time, datetime):
+            update_time = datetime.isoformat(self.update_time)
+        else:
+            update_time = self.create_time
+
+        if self.create_time is None and self.update_time is None:
+            return{
+                "id": self.id,
+                "user_id": self.user_id,
+            }
```
Comment (iLtc) [inline service/models.py:108]: I think we should include `create_time` and `update_time` but set it as `None`.

File: service/models.py [hunk 4]
```diff
@@ -124,32 +178,69 @@ class ShopcartItem(db.Model):
...
+
+        if self.create_time is None and self.update_time is None:
+            return{
+                "id": self.id,
+                "sid": self.sid,
+                "sku": self.sku,
+                "name": self.name,
+                "price": self.price,
+                "amount": self.amount,
+            }
```
Comment (iLtc) [inline service/models.py:233]: Same here. I think we should include `create_time` and `update_time` but set it as `None`.

File: service/routes.py [hunk 5]
```diff
@@ -46,6 +51,51 @@ def index():
...
+@app.route("/shopcarts", methods=["POST"])
+def create_shopcarts():
+    """
+    Creates a Shopcart
+    This endpoint will create a Shopcart based the data in the body that is posted
+    """
+    app.logger.info("Request to create a shopcart")
+    check_content_type("application/json")
+    shopcart = Shopcart()
+    shopcart.deserialize(request.get_json())
```
Comment (iLtc) [inline service/routes.py:66]: If the client sends a `create_time`, do we accept that?

File: service/routes.py [hunk 6]
```diff
@@ -46,6 +51,51 @@ def index():
...
+@app.route("/shopcartitems", methods=["POST"])
+def create_shopcart_items():
+    """
+    Creates a ShopcartItem
+    This endpoint will create a ShopcartItem based the data in the body that is posted
+    """
+    app.logger.info("Request to create a shopcart item")
+    check_content_type("application/json")
+    shopcart_item = ShopcartItem()
+    shopcart_item.deserialize(request.get_json())
```
Comment (iLtc) [inline service/routes.py:89]: Do we need to check if `sid` exists?

**Review by iLtc — CHANGES_REQUESTED** [review]

File: service/models.py (hunk 2 above)
Comment (KHafner) [reply service/models.py:87]: Yeah it wasn't working for me, I thought maybe in future could go back and try and fix this, but didn't want to hold everything up.

File: service/models.py (hunk 4 above)
Comment (KHafner) [reply service/models.py:233]: yeah you're right I'll change it

File: service/routes.py (hunk 6 above)
Comment (KHafner) [reply service/routes.py:89]: wouldn't the deserialize function throw a DataValidationError exception though? I can have it raise an error as well if sid is None

Comment (iLtc) [reply service/routes.py:164]: Sorry, I mean the shopcart that the `sid` refers to. Do we need to check if that shopcart exists?

Comment (KHafner) [reply service/routes.py:164]: well can't do that until the get functionality is added, could add as another issue

File: service/routes.py (hunk 5 above)
Comment (KHafner) [reply service/routes.py:140]: So I'll change the Create function to always use the current time

File: service/routes.py (hunk 6 above)
Comment (iLtc) [reply service/routes.py:164]: Okay, thanks!

**Review by iLtc — APPROVED** [review]
Thanks!

**Review by jules203 — APPROVED** [review]
Great work!

**Comment by nonococoleo** [comment]
should create shopcart function always require an user id, since every shopcart should have an owener?

**Pull request merged**